cloudevents/generic/baseclient.go

```diff
@@ -0,0 +1,209 @@
+package generic
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/flowcontrol"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/clock"
+
+	"open-cluster-management.io/api/cloudevents/generic/options"
+)
+
+const (
+	restartReceiverSignal = iota
+	stopReceiverSignal
+)
+
+type receiveFn func(ctx context.Context, evt cloudevents.Event)
+
+type baseClient struct {
+	sync.RWMutex
+	cloudEventsOptions     options.CloudEventsOptions
+	cloudEventsClient      cloudevents.Client
+	cloudEventsRateLimiter flowcontrol.RateLimiter
+	receiverChan           chan int
+}
+
+func (c *baseClient) connect(ctx context.Context) error {
+	var err error
+	c.cloudEventsClient, err = c.cloudEventsOptions.Client(ctx)
+	if err != nil {
+		return err
+	}
+
+	// start a go routine to handle cloudevents client connection errors
+	go func() {
+		var err error
+
+		// the reconnect backoff will stop at [1,5) min interval. If we don't backoff for 10min, we reset the backoff.
+		connBackoffManager := wait.NewExponentialBackoffManager(5*time.Second, 1*time.Minute, 10*time.Minute, 5.0, 1.0, &clock.RealClock{})
+		cloudEventsClient := c.cloudEventsClient
+
+		for {
+			if cloudEventsClient == nil {
+				klog.V(4).Infof("reconnecting the cloudevents client")
+				cloudEventsClient, err = c.cloudEventsOptions.Client(ctx)
+				// TODO enhance the cloudevents SKD to avoid wrapping the error type to distinguish the net connection
```

Comment:
The reference: https://github.com/eclipse/paho.golang/issues/26 

I will try to add the https://github.com/eclipse/paho.golang/blob/master/autopaho/auto.go code to the mqtt protocol of cloudevents